Add Email Verification

theck13 opened the pull request:
### Fix
Add email verification for users to review their account and verify their email address.  Users can also change their email address if necessary.  See the screenshots below for illustration.

![add_email_verification](https://user-images.githubusercontent.com/3827611/104940985-bbaf4580-596f-11eb-8db6-21c16e01b1a2.png)

### Test
0. Remove app from recents/overview list if present.
1. Launch app.
2. Notice ***Review Your Account*** interface is shown.
3. Tap ***X*** or back system navigation button.
4. Notice ***Review Your Account*** interface is dismissed.
5. Tap recents/overview system navigation button.
6. Tap app in recents/overview list.
7. Notice ***Review Your Account*** interface is not shown.
8. Tap recents/overview system navigation button.
9. Remove app from recents/overview list if present.
10. Launch app.
11. Notice ***Review Your Account*** interface is shown.
12. Tap ***Change Email*** button.
13. Notice web browser is shown with Simplenote login page.
14. Tap back system navigation button.
15. Tap ***Confirm*** button.
16. Notice ***Verify Your Email*** interface is shown.
17. Tap ***X*** or back system navigation button.
18. Notice ***Verify Your Email*** interface is dismissed.
19. Notice verification email is sent to email address shown.
20. Tap link in verification email.
21. Notice web browser is shown with ***Email Verified*** page.
22. Tap ***Open App*** button.
23. Notice app is shown.

### Review
Only one developer and one designer are required to review these changes, but anyone can perform the review.

### Release
`RELEASE-NOTES.txt` was updated in f088432d with:
> Added reviewing, verifying, and changing account email address

theck13 committed:
Add full-screen dialog fragment class

theck13 committed:
Add review account verify email full-screen dialog fragment class

theck13 committed:
Add app log statements to review account verify email

theck13 committed:
Add account bucket class

theck13 committed:
Add checking and showing review account verify email to application class

theck13 committed:
Remove unused method suppression warning from application class

theck13 committed:
Add account bucket to sync worker

theck13 committed:
Add analytics tracking for email verification

theck13 committed:
Add description to release notes for adding email verification

theck13 committed:
Update early exit when checking review account verify email in application class

theck13 committed:
Update string resource for subtitle in verify email full-screen dialog

theck13 committed:
Update checking confirmed account from status to token for email verification

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/models/Account.java:
```diff
@@ -20,13 +21,16 @@ private Account(String key, JSONObject properties) {
             return false;
+        } else if (((String) token).split(":", 2).length > 0) {
+            String emailFromToken = ((String) token).split(":", 2)[0];
+            return emailFromToken.equals(email);
```
would it be helpful to create a separate parsing function for the verification token and then compare the meaningful output here?

```java
class EmailVerificationToken {
    public String email;
    public Calendar verifiedAt;

    public boolean isVerified() { … }
    public boolean isExpired() { … }

    public static EmailVerificationToken fromString(String token) { … }
}


EmailVerificationToken token = EmailVerificationToken.fromString(getProperty(FIELD_EMAIL_VERIFICATION_TOKEN));

return token.isVerified() && token.email.equals(email);
```

theck13 replied:
I thought about making a `Verification` class, but I chose not to since the verification is simple right now.  If we end up adding more to verification like parsing with a public key, then I was going to move everything to a `Verification` class.

Your comment has the `isExpired` method.  Are we planning to expire verification?  If so, what's the logic for that method?

dmsnell replied:
> since the verification is simple right now

it was actually the non-obvious parsing code that scared me. having an explanatory comment at a minimum would make it clearer, rather than only showing implicit expectations based on `:`

```java
// expect tokens in the form [email]:[epoch time in seconds as number]:[verification hash]
// e.g. "simplenote@example.com:193478124:akjn3v9z8ja3jasdf=="
```

of course at this point having a class where the logic is contained seems like it'd be similar in the amount of work but more semantically separated and defined, and less likely to result in duplicated parsing of the token in slightly-different ways around the app

> Are we planning to expire verification?

It was my impression this will happen at some point, such as "It's been a year since you confirmed your email account, can you do that again?"

theck13 replied:
I added comments to the methods in 8b97c9c.

theck13 committed:
Add email address to checking sent email for email verification

theck13 committed:
Add comments to confirmed account and sent email methods in account class

theck13 committed:
Rename method from confirmed account to verified email in account class

theck13 committed:
Update settings redirect url in review account or verify email class

theck13 committed:
Update comment for sent email method in account class

SylvesterWilmott commented:
The UI looks good so far, just a couple of points:

- In dark mode, are we able to eliminate the apparent top app bar so all we have is the "x" up top?
- It looks like the buttons are spaced differently than what we have on the "Sign up" and "Log in" buttons. Can we use the same spacing as we have there?

theck13 commented:
> In dark mode, are we able to eliminate the apparent top app bar so all we have is the "x" up top?

The **_Review Your Account_**/**_Verify Your Email_** interface inherits from a full-screen dialog class, which is using the same styling as other toolbars in the app.  I would be reluctant to change that style since it will affect any future uses of a full-screen dialog.  That will be more of an issue once styles are enabled.  See the screenshots below for illustration.

![add_email_verification_toolbar](https://user-images.githubusercontent.com/3827611/105218863-ee457380-5b12-11eb-9319-c6496a5e6a63.png)

> It looks like the buttons are spaced differently than what we have on the "Sign up" and "Log in" buttons. Can we use the same spacing as we have there?

The spacing was taken from the design in Zeplin, which has 16dp between the buttons.  The buttons for login and signup have no spacing.  See the screenshots below for illustration.

![add_email_verification_butons](https://user-images.githubusercontent.com/3827611/105216977-8130de80-5b10-11eb-8307-f699e66028ea.png)

We can remove the 16dp of spacing if you would like.

SylvesterWilmott commented:
>The Review Your Account/Verify Your Email interface inherits from a full-screen dialog class, which is using the same styling as other toolbars in the app.

Not an issue. Instead, could we add some more space between the content of the screens and the top app bar? Like an additional 8dp.

>We can remove the 16dp of spacing if you would like.

Let's remove it.

theck13 committed:
Add margin above image in review account verify email fragment layout

theck13 committed:
Remove margin between buttons in review account verify email fragment layout

theck13 commented:
I added 8dp margin above the image in 465f9b4f.  I removed 16dp margin between the buttons in 4579a5d2.  See the screenshots below for illustration.

![add_email_verification_spacing](https://user-images.githubusercontent.com/3827611/105224772-d83bb100-5b1a-11eb-820c-1478405abef4.png)

SylvesterWilmott reviewed (approved):
Design looks good! Thanks for the changes @theck13

theck13 committed:
Update parsing in verified email method of account class

theck13 committed:
Update parsing in sent email method of account class

theck13 committed:
Add null check for get arguments in review account verify email fragment

theck13 committed:
Update urls in review account verify email fragment class for testing

theck13 committed:
Revert "Update urls in review account verify email fragment class for testing"

This reverts commit acd3bad3ba70eb6a5d9336973770bc28c558b985.

0nko commented:
I'm not sure if it's related to the PR, but I keep getting this exception (Android 8.1 on the emulator) after logging in using WP.com:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.automattic.simplenote.debug/com.automattic.simplenote.SimplenoteAuthenticationActivity}: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment com.simperium.android.LoginBottomSheetDialogFragment: could not find Fragment constructor
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment com.simperium.android.LoginBottomSheetDialogFragment: could not find Fragment constructor
        at androidx.fragment.app.Fragment.instantiate(Fragment.java:538)
        at androidx.fragment.app.FragmentContainer.instantiate(FragmentContainer.java:57)
        at androidx.fragment.app.FragmentManagerImpl$6.instantiate(FragmentManagerImpl.java:2851)
        at androidx.fragment.app.FragmentState.instantiate(FragmentState.java:85)
        at androidx.fragment.app.FragmentManagerImpl.restoreSaveState(FragmentManagerImpl.java:2488)
        at androidx.fragment.app.FragmentController.restoreSaveState(FragmentController.java:195)
        at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:288)
        at androidx.appcompat.app.AppCompatActivity.onCreate(AppCompatActivity.java:106)
        at com.simperium.android.AuthenticationActivity.onCreate(AuthenticationActivity.java:20)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
     Caused by: java.lang.NoSuchMethodException: <init> []
        at java.lang.Class.getConstructor0(Class.java:2320)
        at java.lang.Class.getConstructor(Class.java:1725)
        at androidx.fragment.app.Fragment.instantiate(Fragment.java:523)
        at androidx.fragment.app.FragmentContainer.instantiate(FragmentContainer.java:57) 
        at androidx.fragment.app.FragmentManagerImpl$6.instantiate(FragmentManagerImpl.java:2851) 
        at androidx.fragment.app.FragmentState.instantiate(FragmentState.java:85) 
        at androidx.fragment.app.FragmentManagerImpl.restoreSaveState(FragmentManagerImpl.java:2488) 
        at androidx.fragment.app.FragmentController.restoreSaveState(FragmentController.java:195) 
        at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:288) 
        at androidx.appcompat.app.AppCompatActivity.onCreate(AppCompatActivity.java:106) 
        at com.simperium.android.AuthenticationActivity.onCreate(AuthenticationActivity.java:20) 
        at android.app.Activity.performCreate(Activity.java:7009) 
        at android.app.Activity.performCreate(Activity.java:7000) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

0nko commented:
Logging in using an email was successful, but when I log out I'm getting the following exception:

```
2021-01-21 17:40:21.311 6219-6219/com.automattic.simplenote.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.automattic.simplenote.debug, PID: 6219
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.automattic.simplenote.debug/com.automattic.simplenote.NotesActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2778)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.equals(java.lang.Object)' on a null object reference
        at com.automattic.simplenote.models.Account.hasVerifiedEmail(Account.java:43)
        at com.automattic.simplenote.Simplenote.checkReviewAccountOrVerifyEmail(Simplenote.java:237)
        at com.automattic.simplenote.Simplenote.access$800(Simplenote.java:58)
        at com.automattic.simplenote.Simplenote$ApplicationLifecycleMonitor.onActivityCreated(Simplenote.java:381)
        at android.app.Application.dispatchActivityCreated(Application.java:197)
        at android.app.Activity.onCreate(Activity.java:1015)
        at androidx.core.app.ComponentActivity.onCreate(ComponentActivity.java:81)
        at androidx.activity.ComponentActivity.onCreate(ComponentActivity.java:149)
        at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:313)
        at androidx.appcompat.app.AppCompatActivity.onCreate(AppCompatActivity.java:106)
        at com.automattic.simplenote.ThemedAppCompatActivity.onCreate(ThemedAppCompatActivity.java:25)
        at com.automattic.simplenote.NotesActivity.onCreate(NotesActivity.java:191)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6494) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807) 
```

Then it's thrown every time the app is opened. The only way to get rid of it is to clear the data.

theck13 commented:
`LoginBottomSheetDialogFragment` wasn't modified in this pull request.  It's actually part of the Simperium Android library, which wasn't changed.  That might be related to the logout issue you're seeing as well.  We should definitely track that and fix it, but I don't believe it's due to these changes.  If you can reproduce it consistently, can you provide the device make/model (real or emulated), Android version, and steps to reproduce it?

theck13 committed:
Update parsing in sent email method of account class

theck13 commented:
@0nko, the backend changed.  I updated the `Account.hasSentEmail` method for those changes in a0deb2f.

0nko reviewed (commented):
I've tested the whole flow before the last change and everything worked as expected. After pulling the latest commits, I'm not getting any confirmation emails.

theck13 committed:
Update urls in review account verify email fragment class for testing

theck13 commented:
I reverted the URL's for testing prematurely.  Sorry about that.  I reverted the revert in cf592a5.  Let me know if you don't get the emails after pulling the latest.

0nko reviewed (approved):
Looks good now 👍

theck13 committed:
Revert "Update urls in review account verify email fragment class for testing"

This reverts commit cf592a5335617585a35328ae145d52696c8a25fd.

theck13 commented:
Thanks, @0nko!  I updated the URL's to their final value in 36a1609, which won't be testable until the backend changes are published, but we can merge this whenever you're ready.

0nko commented:
:shipit:
